### DISCUSSION TITLE
Unique "friendly name" for automatic Zigbee USB adapter discovery in Home Assistant ZHA using dongle vendor/product IDs and description?

### DISCUSSION THREAD

**Discussion by Hedda** [zigpy/zigpy#808] [General]
FYI, Home Assistant developer bdraco has just now added support for automatic USB adapter discovery to [Home Assistant's ZHA integration](https://www.home-assistant.io/integrations/zha/):

https://github.com/home-assistant/core/pull/54904

https://github.com/home-assistant/core/pull/54935

https://github.com/home-assistant/core/pull/54986

If you got any zigpy compatible Zigbee USB adapters which unique Vendor ID (vid) / Product ID (pid) identifier that has not yet been added, please consider submitting pull requests with "friendly name" as description for it to the ZHA component manifest in Home Assistant core. 

Note! As per discussions [#55225](https://github.com/home-assistant/core/issues/55225), [#55236](https://github.com/home-assistant/core/pull/55236) and [#1039](https://github.com/home-assistant/developers.home-assistant/pull/1039), devices need to have unique description as generic + VID and PID are not enough.

https://www.home-assistant.io/integrations/zha/#discovery-via-usb-or-zeroconf

https://github.com/home-assistant/core/blob/dev/homeassistant/components/zha/manifest.json

```
  "usb": [
   {"vid":"10C4","pid":"EA60","description":"*2652*","known_devices":["slae.sh cc2652rb stick"]},
   {"vid":"1CF1","pid":"0030","description":"*conbee*","known_devices":["Conbee II"]},
   {"vid":"10C4","pid":"8A2A","description":"*zigbee*","known_devices":["Nortek HUSBZB-1"]}
  ],
```

https://github.com/home-assistant/core/blob/dev/homeassistant/generated/usb.py

```
USB = [
    {
        "domain": "zha",
        "vid": "10C4",
        "pid": "EA60",
        "description": "*2652*"
    },
    {
        "domain": "zha",
        "vid": "1CF1",
        "pid": "0030",
        "description": "*conbee*"
    },
    {
        "domain": "zha",
        "vid": "10C4",
        "pid": "8A2A",
        "description": "*zigbee*"
    }
```

PS: If follow links you notice this also applies similarly to discovering Z-Wave USB adapter for the [Z-Wave JS component integration](https://www.home-assistant.io/integrations/zwave_js/):

https://www.home-assistant.io/integrations/zwave_js/#discovery-via-usb

https://github.com/home-assistant/core/blob/dev/homeassistant/components/zwave_js/manifest.json

```
  "usb": [
    {"vid":"0658","pid":"0200","known_devices":["Aeotec Z-Stick Gen5+", "Z-WaveMe UZB"]},
    {"vid":"10C4","pid":"8A2A","description":"*z-wave*","known_devices":["Nortek HUSBZB-1"]},
    {"vid":"10C4","pid":"EA60","known_devices":["Aeotec Z-Stick 7", "Silicon Labs UZB-7", "Zooz ZST10 700"]}
  ]
```

https://github.com/home-assistant/core/blob/dev/homeassistant/generated/usb.py

```
{
        "domain": "zwave_js",
        "vid": "0658",
        "pid": "0200"
    },
    {
        "domain": "zwave_js",
        "vid": "10C4",
        "pid": "8A2A",
        "description": "*z-wave*"
    },
    {
        "domain": "zwave_js",
        "vid": "10C4",
        "pid": "EA60"
    }
```

**Comment by Hedda** [comment]
> Note! As per discussions [#55225](https://github.com/home-assistant/core/issues/55225), [#55236](https://github.com/home-assistant/core/pull/55236) and [#1039](https://github.com/home-assistant/developers.home-assistant/pull/1039), devices need to have unique description as generic + VID and PID are not enough.

ITead Zigbee 3.0 USB dongle (HW version “V1.3 2020-12-07”) got VID 1A86 and PID 7523 but that does not look to be unique as seems to be used by the WCH CH340/CH340 serial-to-USB converter/bridge chips in all devices?

Hardware IDs:
```
USB\VID_1A86&PID_7523&REV_0264
```
Device Name Description:
```
USB Serial

```

**Comment by Hedda** [comment]
Tip! Hobbyist with native USB or serial-to-USB chip with writable EEPROM can claim your own unique PID even for DIY projects for free here: 

https://pid.codes/

https://pid.codes/howto/

"_You can get a PID by forking the pid.codes repo, claiming a PID, and sending a pull request. Once that’s accepted, that PID is yours forever._"

https://hackaday.com/2015/04/03/usb-pids-for-all/

PS: I understand that FT231 chip by FTDI or CP2102N by Silicon Labs / Silabs is recommended for such features (and better support).

https://ftdichip.com/products/ft231xs/
https://ftdichip.com/products/ft231xq/

https://www.silabs.com/interface/usb-bridges/classic/device.cp2102
https://www.silabs.com/interface/usb-bridges/classic/device.cp2104
https://www.silabs.com/developers/usb-to-uart-bridge-vcp-drivers
https://www.silabs.com/documents/public/data-sheets/cp2102n-datasheet.pdf
https://www.silabs.com/documents/public/application-notes/an197.pdf
https://www.silabs.com/documents/public/application-notes/an978-cp210x-usb-to-uart-api-specification.pdf
https://www.silabs.com/documents/public/application-notes/AN571.pdf

**Reply by tube0013** [reply]
This shouldn't be done as drivers will not match. The better option is to use the Name and Serial string. This is how the husbzb is detected. 

The cheap ch340 chips do not have a programmable eeprom though. (Aside from the CH340B).

**Reply by mercenaruss** [reply]
CH340B is a good alternative. All another are too expensive and not always available.

**Reply by Hedda** [reply]
> This shouldn't be done as drivers will not match. The better option is to use the Name and Serial string. This is how the husbzb is detected.

Correct, Home Assistant and ZHA only use descriptor in combination with VID and PID to auto detect and identify unique devices.

You do not want to change VID and PID as the operating system uses those two to select which device driver to use.

This way multiple devices from different manufacturers can have same VID and PID while still being detected by app as unique.

Note! If you do change the VID or PID you must provide the customer with your own custom driver for your specific device.

**Comment by Hedda** [comment]
FYI, I read that the upcoming ZZHP and ZZHP-Lite (Zig-a-zig-ah v2) adapters by Electrolama will probably use FT231 or CP2102/P2104 which will allow them to write a customized string onto the EEPROM of the USB-to-Serial chip (such as for example "Electrolama ZZHP v1.0.0") and more stuff.

Should make it more plug-and-play friendly and easier for different home automation software to automatically discover and configure without end-user interactions. 

Electrolama's ZZH (Zig-a-zig-ah v1) adapter has been removed for this reason from HA's USB discovery -> https://github.com/home-assistant/core/pull/55236

**Comment by Hedda** [comment]
FYI, Home Assistant developers elected to not whitelist "TI CC2531 USB CDC" ZHA discovery due to its old hardware:

https://github.com/home-assistant/core/pull/55298

```
{"vid":"0451","pid":"16A8","known_devices":["TI CC2531 USB CDC"]}
```

PS: Zigbee2MQTT now added "not recommended" notes to CC2530 and CC2531 in their official supported adapters list:

https://www.zigbee2mqtt.io/information/supported_adapters

**Comment by Hedda** [comment]
FYI, Home Assistant and ZHA will soon also detect tubeszb USB connected devices that uses CH340B serial chip with a custom descriptor:

https://github.com/home-assistant/core/pull/56719

https://github.com/home-assistant/core/pull/56719/commits/c3c0e343aacde82d59878619594199278ee67b01

  homeassistant/components/zha/manifest.json 

```
   {"vid":"1A86","pid":"7523","description":"*tubeszb*","known_devices":["TubesZB Coordinator"]},
```

homeassistant/generated/usb.py 

```
    {
        "domain": "zha",
        "vid": "1A86",
        "pid": "7523",
        "description": "*tubeszb*"
    },
```

**Reply by mercenaruss** [reply]
@Hedda i managed change description on CH340B , now all what remains to submit a PR to core. Thank you for assistance.
